packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts

```
@@ -0,0 +1,98 @@
+import { PricesObj } from "../../types";
+import {
+  BalancerSDK,
+  BalancerSdkConfig,
+  Network,
+  PoolWithMethods,
+} from "@balancer-labs/sdk";
+import { config } from "../../config";
+import { BaseFetcher } from "../BaseFetcher";
+import { getLastPrice } from "../../db/local-db";
+import balancerPairs from "./balancer-pairs.json";
+
+const balancerConfig: BalancerSdkConfig = {
+  network: Network.MAINNET,
+  rpcUrl: config.ethMainRpcUrl as string,
+};
+
+interface SpotPrice {
+  id: string;
+  price: number;
+}
+
+export class BalancerFetcher extends BaseFetcher {
+  private balancer: BalancerSDK;
+
+  constructor(name: string, protected readonly baseTokenSymbol: string) {
+    super(name);
+    this.balancer = new BalancerSDK(balancerConfig);
+  }
+
+  async fetchData(ids: string[]): Promise<any> {
+    const spotPrices: SpotPrice[] = [];
+
+    const pairIds = this.getPairIdsForAssetIds(ids);
+
+    const pairedTokenPrice = await this.getPairedTokenPrice();
+    for (const pairId of pairIds) {
+      const priceResult = await this.calculatePrice(pairId, pairedTokenPrice);
+      if (!priceResult) {
+        throw new Error(`Could not get pool with id ${pairId}`);
+      }
+      const { id, price } = priceResult;
+      spotPrices.push({ id, price });
+    }
+    return spotPrices;
+  }
+
+  protected async calculatePrice(
+    pairId: string,
+    pairedTokenPrice: number
+  ): Promise<SpotPrice | null> {
+    const pool = await this.balancer.pools.find(pairId);
+    if (pool) {
+      const spotPrice = pool.calcSpotPrice(
+        pool.tokens[0].address,
+        pool.tokens[1].address
+      );
+      const price = pairedTokenPrice / Number(spotPrice);
+      return { id: this.getSymbol(pool), price };
+    }
+    return null;
```

Let's just throw an error if pool is not found